gk/main.c

```diff
@@ -552,6 +552,72 @@ gk_setup_rss(struct gk_config *gk_conf)
 	return ret;
 }
 
+static void
+gk_arp_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	unsigned int i;
+	struct gk_config *gk_conf = arg;
+	struct gk_lpm *ltbl = gk_conf->lpm_tbl;
+
+	for (i = 0; i < ltbl->num_tunnels; i++) {
+		if (ltbl->tunnels[i].flow.proto == ETHER_TYPE_IPv4
+				&& memcmp(&ltbl->tunnels[i].flow.f.v6.dst,
+				map->ip_be, sizeof(struct in_addr)) == 0) {
+			ether_addr_copy(&map->ha,
+				&ltbl->tunnels[i].nexthop_mac);
+			break;
+		}
+	}
+
+	*pcall_again = true;
+}
+
+static void
+gk_nd_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	unsigned int i;
+	struct gk_config *gk_conf = arg;
+	struct gk_lpm *ltbl = gk_conf->lpm_tbl;
+
+	for (i = 0; i < ltbl->num_tunnels; i++) {
+		if (ltbl->tunnels[i].flow.proto == ETHER_TYPE_IPv6
+				&& memcmp(ltbl->tunnels[i].flow.f.v6.dst,
+				map->ip_be, sizeof(struct in6_addr)) == 0) {
+			ether_addr_copy(&map->ha,
+				&ltbl->tunnels[i].nexthop_mac);
+			break;
+		}
+	}
+
+	*pcall_again = true;
+}
+
+static int
+gk_hold_arp_and_nd(struct gk_config *gk_conf, unsigned int lcore_id)
+{
+	unsigned int i;
+	int ret;
+	struct gk_lpm *ltbl = gk_conf->lpm_tbl;
+
+	for (i = 0; i < ltbl->num_tunnels; i++) {
+		if (ltbl->tunnels[i].flow.proto == ETHER_TYPE_IPv4) {
+			ret = hold_arp(gk_arp_req_cb, (void *)gk_conf, (struct in_addr *)
+				&ltbl->tunnels[i].flow.f.v4.dst, lcore_id);
```

Comment:
Yes to (1) and (2).

If policy here means forwarding to a gateway, yes to the question.

Notice that you can avoid the double lookup having a reference from the grantor entry to the gateway entry.